### PR TITLE
bazel_5: 5.2.0 -> 5.3.2

### PR DESCRIPTION
**Pull request by divanorama**
###### Description of changes

https://blog.bazel.build/2022/08/23/bazel-5.3.html

to supersede https://github.com/NixOS/nixpkgs/pull/187943 and https://github.com/NixOS/nixpkgs/pull/198016

re-reverts https://github.com/NixOS/nixpkgs/pull/199458 with more hash updates: bazel-watcher, jaxlib, tensorflow, envoy. That should be all using `buildBazelPackage` with `bazel_5`. On darwin tensorflow still fails, so replacing hash with explicitly invalid one.

ZHF: #199919 (fixes bazel-watcher on x86_64-darwin)

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [X] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by divanorama** [comment]
https://github.com/ofborg build envoy

**Comment by divanorama** [comment]
@ofborg build tensorflow

**Comment by divanorama** [comment]
@ofborg build python3Packages.jaxlib

File: pkgs/servers/http/envoy/default.nix [hunk 1]
```diff
@@ -83,8 +83,8 @@ buildBazelPackage rec {
 
   fetchAttrs = {
     sha256 = {
-      x86_64-linux = "10f1lcn8pynqcj2hlz100zbpmawvn0f2hwpcw3m9v6v3fcs2l6pr";
-      aarch64-linux = "1na7gna9563mm1y7sy34fh64f1kxz151xn26zigbi9amwcpjbav6";
+      x86_64-linux = "sha256-l/BWUKmLr5m6OJDLMMuXA4w6eRX7jEzNMd8xV0vF+Rc=";
+      aarch64-linux = "sha256-/jYUnuOESJ7FfszRk1QptFpuL/ch0DroHTJXdLaqsCk=";
```
Comment (divanorama) [inline pkgs/servers/http/envoy/default.nix:87]:
```suggestion
      aarch64-linux = "sha256-/L/QC1nOLqLUGoXfNSXYqA+QHJ8euLLXPvfNEpTGBBc=";
```
hydra produces this value (should probably investigate the difference somehow)

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 2]
```diff
@@ -369,13 +369,13 @@ let
...
-        "sha256-KtVReqHL3zxE8TPrqIerSOt59Mgke/ftoFZKMzgX/u8="
+        "sha256-PnPT+9JPrg9PfB9d7N4LHFs/RkESWowlQDPOnsTlGEg="
       else
         if stdenv.isDarwin then
           # FIXME: this checksum is currently wrong, since the tensorflow dependency fetch is broken on darwin
-          "sha256-j2k9Q+k41nq5nP1VjjkkNjXRov1uAda4RCMDMAthjr0="
+          lib.fakeHash
         else
-          "sha256-zH3xNFEU2JR0Ww8bpD4mCiorGtao0WVPP4vklVMgS4A=";
+          "sha256-3nAjP6WH6xhvVyA6ZGZ6ZpsTTUCQohNIiQhVn1eWpr8=";
```
Comment (divanorama) [inline pkgs/development/python-modules/tensorflow/default.nix:378]:
```suggestion
          "sha256-FKz6SRmS5E/BJ1WqIjRiVKGfgedmRG6ip/eAmn6LdcQ=";
```
ofborg value

File: pkgs/development/python-modules/jaxlib/default.nix [hunk 3]
```diff
@@ -235,11 +235,11 @@ let
     fetchAttrs = {
       sha256 =
         if cudaSupport then
-          "sha256-Z9GDWGv+1YFyJjudyshZfeRJsKShoA1kIbNR3h3GxPQ="
+          "sha256-FO3mvxr+c3kWOhIRRdJ0gBOHrPr2LKrtjZvYaPp8Xu0="
         else if stdenv.isDarwin then
-          "sha256-i3wiJHD4+pgTvDMhnYiQo9pdxxKItgYnc4/4wGt2NXM="
+          "sha256-/C6HUTHvRd+3LL4repJWWQq+2UXgO5za4xXZhhNxNrM="
```
Comment (divanorama) [inline pkgs/development/python-modules/jaxlib/default.nix:240]:
```suggestion
          "sha256-xF36ctkzH/LGvKhrcrqVDz5tUqFnYsFaOzjf3aJUTtU="
```
from ofborg

**Comment by divanorama** [comment]
@ofborg build envoy
@ofborg build python3Packages.jaxlib
@ofborg build python3Packages.tensorflow

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 4]
```diff
@@ -369,13 +369,13 @@ let
...
-        "sha256-KtVReqHL3zxE8TPrqIerSOt59Mgke/ftoFZKMzgX/u8="
+        "sha256-PnPT+9JPrg9PfB9d7N4LHFs/RkESWowlQDPOnsTlGEg="
       else
         if stdenv.isDarwin then
           # FIXME: this checksum is currently wrong, since the tensorflow dependency fetch is broken on darwin
-          "sha256-j2k9Q+k41nq5nP1VjjkkNjXRov1uAda4RCMDMAthjr0="
+          lib.fakeHash
         else
-          "sha256-zH3xNFEU2JR0Ww8bpD4mCiorGtao0WVPP4vklVMgS4A=";
+          "sha256-FKz6SRmS5E/BJ1WqIjRiVKGfgedmRG6ip/eAmn6LdcQ=";
```
Comment (divanorama) [inline pkgs/development/python-modules/tensorflow/default.nix:378]:
```suggestion
          "sha256-3nAjP6WH6xhvVyA6ZGZ6ZpsTTUCQohNIiQhVn1eWpr8=";
```

**Comment by divanorama** [comment]
@ofborg build envoy
@ofborg build bazel-watcher
@ofborg build python3Packages.jaxlib
@ofborg build python3Packages.tensorflow
@ofborg build python3Packages.jaxlibWithCuda
@ofborg build python3Packages.tensorflowWithCuda

**Comment by divanorama** [comment]
@ofborg build envoy
@ofborg build bazel-watcher
@ofborg build verible
@ofborg build python3Packages.jaxlib
@ofborg build python3Packages.ray
@ofborg build python3Packages.tensorflow
@ofborg build python3Packages.jaxlibWithCuda
@ofborg build python3Packages.tensorflowWithCuda

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -16367,7 +16367,7 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) CoreServices;
   };
 
-  buildBazelPackage = callPackage ../build-support/build-bazel-package { };
+  buildBazelPackage = (if stdenv.isDarwin then darwin.apple_sdk_11_0.callPackage else callPackage) ../build-support/build-bazel-package { };
```
Comment (divanorama) [inline pkgs/top-level/all-packages.nix:16370]: borrowed this fix from https://github.com/NixOS/nixpkgs/pull/198701#discussion_r1009044255
It fixes `bazel-watcher` on x86_64-darwin, but only if placed here, changing `callPackage` on `bazel-watcher` itself doesn't help

**Comment by divanorama** [comment]
@ofborg build envoy
@ofborg build bazel-watcher
@ofborg build verible
@ofborg build python3Packages.jaxlib
@ofborg build python3Packages.ray
@ofborg build python3Packages.tensorflow
@ofborg build python3Packages.jaxlibWithCuda
@ofborg build python3Packages.tensorflowWithCuda

**Comment by divanorama** [comment]
python3Packages.jaxlib and python3Packages.tensorflow on x86_64-darwin failures seem to be unrelated
https://github.com/bazelbuild/bazel/issues/5127
https://github.com/bazelbuild/bazel/issues/15359

bazel, bazel.passthru.tests on aarch64-darwin is bazel_3 failing on installCheckPhase, should also be unrelated

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:16370]: This is wrong in the PR and shouldn't be used like this.

```suggestion
  buildBazelPackage = darwin.apple_sdk_11_0.callPackage ../build-support/build-bazel-package { };
```

Comment (divanorama) [reply pkgs/top-level/all-packages.nix:16370]: Looked more logical that for -linux it shouldn't be using darwin.*.callPackage, but will apply :+1:

Comment (divanorama) [reply pkgs/top-level/all-packages.nix:16370]: updated and squashed commits together

**Comment by risicle** [comment]
`envoy` fails for me on aarch64-linux

```
error: hash mismatch in fixed-output derivation '/nix/store/4x5zgyw71nbph204nzxj2fab1c7j53g4-envoy-1.23.1-deps.tar.gz.drv':
         specified: sha256-/L/QC1nOLqLUGoXfNSXYqA+QHJ8euLLXPvfNEpTGBBc=
            got:    sha256-AkNQ5+HVliSQAB5hmrW1eADKfW/GQhxZZPo3S+eISdo=
```

File: pkgs/servers/http/envoy/default.nix [hunk 6]
```diff
@@ -83,8 +83,8 @@ buildBazelPackage rec {
 
   fetchAttrs = {
     sha256 = {
-      x86_64-linux = "10f1lcn8pynqcj2hlz100zbpmawvn0f2hwpcw3m9v6v3fcs2l6pr";
-      aarch64-linux = "1na7gna9563mm1y7sy34fh64f1kxz151xn26zigbi9amwcpjbav6";
+      x86_64-linux = "sha256-l/BWUKmLr5m6OJDLMMuXA4w6eRX7jEzNMd8xV0vF+Rc=";
+      aarch64-linux = "sha256-/L/QC1nOLqLUGoXfNSXYqA+QHJ8euLLXPvfNEpTGBBc=";
```
Comment (divanorama) [inline pkgs/servers/http/envoy/default.nix:87]:
```suggestion
      x86_64-linux = lib.fakeHash;
      aarch64-linux = lib.fakeHash;
```
let's debug via ofborg, I also now get different digest locally (maybe after rebases)

**Comment by divanorama** [comment]
@ofborg build envoy

**Comment by risicle** [comment]
Builds for me now, aarch64-linux (possible I just got lucky though)

**Comment by risicle** [comment]
`nixpkgs-review` reveals no new failures, aarch64-linux

**Comment by mweinelt** [comment]
Looks like the build on aarch64-darwin is broken.

**Comment by divanorama** [comment]
https://github.com/ofborg build bazel_5

**Comment by divanorama** [comment]
`bazel` aka `bazel_3` is failing on ofborg on aarc64-darwin, bazel_5 should be fine (works locally at least)

maybe because PR title had `bazel: ` prefix it decided to build bazel_3, requested explicit bazel_5 build now and changed the title

**Comment by divanorama** [comment]
yep, bazel_5 builds on ofborg, bazel_3 failure on aarch64-darwin should be unrelated

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix]: The callPackage is a wrapper that for linux is a noop.

Comment (divanorama) [reply pkgs/top-level/all-packages.nix]: I don't doubt that it currently works either way. But I can imagine there being a darwin.apple_sdk_11_0 wrapper, and say openbsd.something-something or whatever that own act only on their specific systems and usecases. And then which should be used if package needs both tweaks, not sure how to compose them. And not sure how to improve readability for default case, we aren't going to use apple_sdk on linux, but just looking at `darwin.apple_sdk_11_0.callPackage` there could be impression that we would / that it is a darwin-only package.

But anyway, I applied the suggestion and this general improvements stuff goes a little beyond the scope of bazel_5 version bump :slightly_smiling_face:

**Comment by SuperSandro2000** [comment]
Sorry for the delay. Please rebase and then we can merge this.

**Comment by divanorama** [comment]
Rebasing, will need to update some hashes.
Meanwhile 5.4.0 got released, should we skip 5.3.2 or better make it a separate PR?

**Comment by divanorama** [comment]
rebased now :+1:

**Comment by divanorama** [comment]
If straight to 5.4.0 should be better, here's a draft https://github.com/NixOS/nixpkgs/pull/207870

**Pull request merged**